src/openfermion/utils/_bch_expansion.py

```diff
@@ -0,0 +1,157 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""Module to efficiently compute the Baker-Campbell-Hausdorff formula."""
+
+import itertools
+import numpy as np
+from scipy.misc import comb, factorial
+
+from openfermion.utils import commutator
+
+def bch_expand(x, y, order):
+    """Compute log[e^x e^y] using the Baker-Campbell-Hausdorff formula
+    Args:
+        x: An operator for which multiplication and addition are supported.
+            For instance, a QubitOperator, FermionOperator or numpy array.
+        y: The same type as x.
+        order(int): The order to truncate the BCH expansions.
+
+    Returns:
+        z: The truncated BCH operator.
+
+    Raises:
+        ValueError: operator x is not same type as operator y.
+        ValueError: invalid order parameter.
+        ValueError: order exceeds maximum order supported.
+    """
+
+    max_order = 11
+    if (not isinstance(order, int)) or order < 0:
+        raise ValueError('Invalid order parameter.')
+    if order > max_order:
+        raise ValueError('Order exceeds maximum order supported.')
+    if type(x) != type(y):
+        raise ValueError('Operator x is not same type as operator y.')
+
+    z = None
+    terms, coeff = generate_nested_commutator(order)
+    for bin_str, c in zip(terms, coeff):
+        t = bin_str_to_commutator(bin_str, x, y)
+        if z is None:
+            z = t * c
+        else:
+            z += t * c
+
+    # Return.
+    return z
+
+def bin_str_to_commutator(bin_str, x, y):
```

Comment:
All functions (even helper functions) must have at least a short one line description. Ideally, also provide args and returns.

Reply:
Correct me if I'm wrong, but I think this is a fairly time-consuming operation (need to compute one commutator per order) that are called many times (one time for each term in the expansion, and then in recursion) with same parameters (many intermediate results are previously computed terms). Optimization potential here?

Reply:
Yes, I'll optimize it after v0.2. It's fast enough for 10by10 matrices up to order 7-8 so I think I'll leave like this in the current version, because it makes the code way easier to read.

Reply:
That's fine with me.